subprojects/core/src/main/java/org/gradle/internal/classpath/CallInterceptionClosureInstrumentingClassVisitor.java

```diff
@@ -0,0 +1,281 @@
+
+                    _INVOKESTATIC(Type.getType(InstrumentedGroovyMetaClassHelper.class), "addInvocationHooksToEffectivelyInstrumentClosure", getMethodDescriptor(Type.VOID_TYPE, CLOSURE_TYPE));
+                }
+            }
+            return new MethodVisitorScopeImpl(classData.visitor.visitMethod(Opcodes.ACC_PUBLIC, methodData.name, "()V", null, null));
+        }),
+
+        /**
+         * Does not perform any transformations.
+         */
+        DEFAULT(null, null, false, (classData, mv) -> classData.visitor.visitMethod(mv.access, mv.name, mv.descriptor, mv.signature, mv.exceptions));
+
+        public final @Nullable String methodName;
+        public final @Nullable String descriptor;
+        public final boolean generateIfNotPresent;
+        private final BiFunction<ClassData, MethodData, MethodVisitor> methodVisitorFactory;
+
+        @NonNullApi
+        static final class MethodData {
+            public final int access;
+            public final String name;
+            public final String descriptor;
+            public final @Nullable String signature;
+            public final @Nullable String[] exceptions;
+
+            public MethodData(int access, String name, String descriptor, @Nullable String signature, @Nullable String[] exceptions) {
+                this.access = access;
+                this.name = name;
+                this.descriptor = descriptor;
+                this.signature = signature;
+                this.exceptions = exceptions;
+            }
+        }
+
+        @NonNullApi
+        static final class ClassData {
+            public final ClassVisitor visitor;
+            public final String className;
+
+            ClassData(ClassVisitor visitor, String className) {
+                this.visitor = visitor;
+                this.className = className;
+            }
+        }
+
+        MethodInstrumentationStrategy(
+            @Nullable String methodName,
+            @Nullable String descriptor,
+            boolean generateIfNotPresent,
+            BiFunction<ClassData, MethodData, MethodVisitor> methodVisitorFactory
+        ) {
+            this.methodName = methodName;
+            this.descriptor = descriptor;
+            this.generateIfNotPresent = generateIfNotPresent;
+            this.methodVisitorFactory = methodVisitorFactory;
+        }
+    }
+
+    boolean inClosureImplementation = false;
+    String className = null;
+    EnumSet<MethodInstrumentationStrategy> usedStrategies = EnumSet.noneOf(MethodInstrumentationStrategy.class);
+
+    private static final Type CLOSURE_TYPE = getType(Closure.class);
+    private static final String CLOSURE_INTERNAL_NAME = CLOSURE_TYPE.getInternalName();
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+        boolean isClosureImplementation = CLOSURE_INTERNAL_NAME.equals(superName);
```

Comment:
I never encountered "natural" closure implementations that would not be direct subclasses. 
I think we cannot reliable find out what is the closure's body when it's hand-written (e.g. it may have multiple methods called from the body). If it is not written in Groovy, it will have static references that won't need dynamic call interception anyway. So for now the part that is not covered is `Closure` implementations written by hand in Groovy, which is a rather exotic edge case.